mlflow/store/job/abstract_store.py

```diff
@@ -0,0 +1,137 @@
+from typing import Iterator
+
+from mlflow.entities._job import Job
+from mlflow.entities._job_status import JobStatus
+from mlflow.utils.annotations import developer_stable
+
+
+@developer_stable
+class AbstractJobStore:
+    """
+    Abstract class that defines API interfaces for storing Job metadata.
+    """
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self):
+        """
+        Empty constructor. This is deliberately not marked as abstract, else every derived class
+        would be forced to create one.
+        """
+
+    @abstractmethod
+    def create_job(self, function_fullname: str, params: str, timeout: int | None = None) -> Job:
+        """
+        Create a new job with the specified function and parameters.
+
+        Args:
+            function_fullname: The full name of the function to execute
+            params: The job parameters as a string
+            timeout: The job execution timeout in seconds
+
+        Returns:
+            Job entity instance
+        """
+
+    @abstractmethod
+    def start_job(self, job_id: str) -> None:
+        """
+        Start a job by setting its status to RUNNING.
+
+        Args:
+            job_id: The ID of the job to start
+        """
+
+    @abstractmethod
+    def reset_job(self, job_id: str) -> None:
+        """
+        Reset a job by setting its status to PENDING.
+
+        Args:
+            job_id: The ID of the job to re-enqueue.
+        """
+
+    @abstractmethod
+    def finish_job(self, job_id: str, result: str) -> None:
+        """
+        Finish a job by setting its status to DONE and setting the result.
+
+        Args:
+            job_id: The ID of the job to finish
+            result: The job result as a string
+        """
+
+    @abstractmethod
+    def set_job_timeout(self, job_id: str) -> None:
+        """
+        Set a job status to Timeout.
+
+        Args:
+            job_id: The ID of the job
+        """
+
+    @abstractmethod
+    def fail_job(self, job_id: str, error: str) -> None:
+        """
+        Fail a job by setting its status to FAILED and setting the error message.
+
+        Args:
+            job_id: The ID of the job to fail
+            error: The error message as a string
+        """
+
+    @abstractmethod
+    def retry_or_fail_job(self, job_id: str, error: str) -> int | None:
+        """
+        If the job retry_count is less than maximum allowed retry count,
+        increase the retry_count and reset the job to PENDING status,
+        otherwise set the job to FAILED status and fill the job's error field.
+
+        Args:
+            job_id: The ID of the job to fail
+            error: The error message as a string
+
+        Returns:
+            If the job is allowed to retry, returns the retry count,
+            otherwise returns None.
+        """
+
+    @abstractmethod
+    def list_jobs(
+        self,
+        function_fullname: str | None = None,
+        status_list: list[JobStatus] | None = None,
```

Comment:
```suggestion
        statuses: list[JobStatus] | None = None,
```

we can make this plural unless there is a reason we need to use `_list`.